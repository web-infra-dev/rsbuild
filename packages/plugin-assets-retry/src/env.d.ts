declare const RSBUILD_VERSION;
