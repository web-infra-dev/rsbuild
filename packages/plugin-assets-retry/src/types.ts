import type { CrossOrigin } from '@rsbuild/shared';

export type PluginAssetsRetryOptions = {
  max?: number;
  type?: string[];
  test?: string | ((url: string) => boolean);
  domain?: string[];
  crossOrigin?: boolean | CrossOrigin;
  inlineScript?: boolean;
  onFail?: (options: AssetsRetryHookContext) => void;
  onRetry?: (options: AssetsRetryHookContext) => void;
  onSuccess?: (options: AssetsRetryHookContext) => void;
};

export type AssetsRetryHookContext = {
  url: string;
  times: number;
  domain: string;
  tagName: string;
};
