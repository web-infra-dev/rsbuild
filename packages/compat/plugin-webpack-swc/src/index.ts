export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
