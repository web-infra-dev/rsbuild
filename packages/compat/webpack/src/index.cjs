/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
