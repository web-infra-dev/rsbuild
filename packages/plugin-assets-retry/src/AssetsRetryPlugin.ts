import path from 'node:path';
import {
  type Rspack,
  fse,
  generateScriptTag,
  getPublicPathFromCompiler,
  withPublicPath,
} from '@rsbuild/shared';
import type HtmlWebpackPlugin from 'html-webpack-plugin';
import type { PluginAssetsRetryOptions } from './types';

export class AssetsRetryPlugin implements Rspack.RspackPluginInstance {
  readonly name: string;

  readonly distDir: string;

  readonly inlineScript: boolean;

  readonly minify: boolean;

  readonly HtmlPlugin: typeof HtmlWebpackPlugin;

  scriptPath: string;

  readonly #retryOptions: PluginAssetsRetryOptions;

  constructor(
    options: PluginAssetsRetryOptions & {
      distDir: string;
      HtmlPlugin: typeof HtmlWebpackPlugin;
    },
  ) {
    const {
      distDir,
      HtmlPlugin,
      inlineScript = true,
      minify = process.env.NODE_ENV === 'production',
      ...retryOptions
    } = options;
    this.name = 'AssetsRetryPlugin';
    this.#retryOptions = retryOptions;
    this.distDir = distDir;
    this.HtmlPlugin = HtmlPlugin;
    this.inlineScript = inlineScript;
    this.scriptPath = '';
    this.minify = minify;
  }

  async getRetryCode() {
    const { default: serialize } = await import('serialize-javascript');
    const filename = 'initialChunkRetry';
    const runtimeFilePath = path.join(
      __dirname,
      'runtime',
      this.minify ? `${filename}.min.js` : `${filename}.js`,
    );
    const runtimeCode = await fse.readFile(runtimeFilePath, 'utf-8');
    return `(function(){${runtimeCode};init(${serialize(
      this.#retryOptions,
    )});})()`;
  }

  async getScriptPath() {
    if (!this.scriptPath) {
      this.scriptPath = path.posix.join(
        this.distDir,
        `assets-retry.${RSBUILD_VERSION}.js`,
      );
    }
    return this.scriptPath;
  }

  apply(compiler: Rspack.Compiler): void {
    if (!this.inlineScript) {
      compiler.hooks.thisCompilation.tap(
        this.name,
        (compilation: Rspack.Compilation) => {
          compilation.hooks.processAssets.tapPromise(
            {
              name: this.name,
              stage:
                compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_PRE_PROCESS,
            },
            async () => {
              const scriptPath = await this.getScriptPath();
              const code = await this.getRetryCode();
              compilation.emitAsset(
                scriptPath,
                new compiler.webpack.sources.RawSource(code, false),
              );
            },
          );
        },
      );
    }

    compiler.hooks.compilation.tap(this.name, (compilation) => {
      // the behavior of inject/modify tags in afterTemplateExecution hook will not take effect when inject option is false
      this.HtmlPlugin.getHooks(compilation).alterAssetTagGroups.tapPromise(
        this.name,
        async (data) => {
          const scriptTag = generateScriptTag();

          if (this.inlineScript) {
            data.headTags.unshift({
              ...scriptTag,
              innerHTML: await this.getRetryCode(),
            });
          } else {
            const publicPath = getPublicPathFromCompiler(compiler);
            const url = withPublicPath(await this.getScriptPath(), publicPath);
            data.headTags.unshift({
              ...scriptTag,
              attributes: {
                ...scriptTag.attributes,
                src: url,
              },
            });
          }

          return data;
        },
      );
    });
  }
}
