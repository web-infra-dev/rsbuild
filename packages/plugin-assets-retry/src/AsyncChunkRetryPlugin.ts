import fs from 'node:fs';
import path from 'node:path';
import { type Rspack, rspack } from '@rsbuild/core';
import serialize from 'serialize-javascript';
import type { PluginAssetsRetryOptions, RuntimeRetryOptions } from './types';

// https://github.com/web-infra-dev/rspack/pull/5370
function appendWebpackScript(module: any, appendSource: string) {
  try {
    const originSource = module.getGeneratedCode();
    module.getGeneratedCode = () => `${originSource}\n${appendSource}`;
  } catch (err) {
    console.error('Failed to modify Webpack RuntimeModule');
    throw err;
  }
}

function appendRspackScript(
  module: any, // JsRuntimeModule type is not exported by rspack temporarily */
  appendSource: string,
) {
  try {
    const source = module.source.source.toString();
    module.source.source = Buffer.from(`${source}\n${appendSource}`, 'utf-8');
  } catch (err) {
    console.error('Failed to modify Rspack RuntimeModule');
    throw err;
  }
}

function pick<T, U extends keyof T>(obj: T, keys: ReadonlyArray<U>) {
  return keys.reduce(
    (ret, key) => {
      if (obj[key] !== undefined) {
        ret[key] = obj[key];
      }
      return ret;
    },
    {} as Pick<T, U>,
  );
}

class AsyncChunkRetryPlugin implements Rspack.RspackPluginInstance {
  readonly name = 'ASYNC_CHUNK_RETRY_PLUGIN';
  readonly options: PluginAssetsRetryOptions & { isRspack: boolean };
  readonly runtimeOptions: RuntimeRetryOptions;

  constructor(options: PluginAssetsRetryOptions & { isRspack: boolean }) {
    this.options = options;
    this.runtimeOptions = pick(options, [
      'domain',
      'max',
      'onRetry',
      'onSuccess',
      'onFail',
      'addQuery',
      'test',
    ]);
  }

  getRawRuntimeRetryCode() {
    const { RuntimeGlobals } = rspack;
    const filename = 'asyncChunkRetry';
    const runtimeFilePath = path.join(
      __dirname,
      'runtime',
      this.options.minify ? `${filename}.min.js` : `${filename}.js`,
    );
    const rawText = fs.readFileSync(runtimeFilePath, 'utf-8');

    return rawText
      .replaceAll('__RUNTIME_GLOBALS_REQUIRE__', RuntimeGlobals.require)
      .replaceAll(
        '__RUNTIME_GLOBALS_ENSURE_CHUNK__',
        RuntimeGlobals.ensureChunk,
      )
      .replaceAll(
        '__RUNTIME_GLOBALS_GET_CHUNK_SCRIPT_FILENAME__',
        RuntimeGlobals.getChunkScriptFilename,
      )
      .replaceAll('__RUNTIME_GLOBALS_PUBLIC_PATH__', RuntimeGlobals.publicPath)
      .replaceAll('__RUNTIME_GLOBALS_LOAD_SCRIPT__', RuntimeGlobals.loadScript)
      .replaceAll('__RETRY_OPTIONS__', serialize(this.runtimeOptions));
  }

  apply(compiler: Rspack.Compiler) {
    compiler.hooks.thisCompilation.tap(this.name, (compilation) => {
      compilation.hooks.runtimeModule.tap(this.name, (module) => {
        const { isRspack } = this.options;
        const constructorName = isRspack
          ? module.constructorName
          : module.constructor?.name;
        const isPublicPathModule =
          module.name === 'publicPath' ||
          constructorName === 'PublicPathRuntimeModule';
        if (!isPublicPathModule) {
          return;
        }

        const runtimeCode = this.getRawRuntimeRetryCode();

        // Rspack currently does not have module.addRuntimeModule on the js side,
        // so we insert our runtime code after PublicPathRuntimeModule.
        if (isRspack) {
          appendRspackScript(module, runtimeCode);
        } else {
          appendWebpackScript(module, runtimeCode);
        }
      });
    });
  }
}

export { AsyncChunkRetryPlugin };
