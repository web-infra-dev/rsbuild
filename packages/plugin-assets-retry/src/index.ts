import type { RsbuildPlugin } from '@rsbuild/core';
import type { PluginAssetsRetryOptions } from './types';

export type { PluginAssetsRetryOptions };

export const PLUGIN_ASSETS_RETRY_NAME = 'rsbuild:assets-retry';

export const pluginAssetsRetry = (
  options: PluginAssetsRetryOptions = {},
): RsbuildPlugin => ({
  name: PLUGIN_ASSETS_RETRY_NAME,
  setup(api) {
    api.modifyBundlerChain(
      async (chain, { CHAIN_ID, HtmlPlugin, isProd, environment }) => {
        const config = api.getNormalizedConfig();

        const htmlPaths = api.getHTMLPaths({ environment });
        if (!options || Object.keys(htmlPaths).length === 0) {
          return;
        }

        const { AssetsRetryPlugin } = await import('./AssetsRetryPlugin');
        const { AsyncChunkRetryPlugin } = await import(
          './AsyncChunkRetryPlugin'
        );
        const distDir = config.output.distPath.js;

        // options.crossOrigin should be same as html.crossorigin by default
        if (options.crossOrigin === undefined) {
          options.crossOrigin = config.html.crossorigin;
        }

        if (options.minify === undefined) {
          const minify =
            typeof config.output.minify === 'boolean'
              ? config.output.minify
              : config.output.minify?.js;
          options.minify = minify && isProd;
        }

        chain.plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY).use(AssetsRetryPlugin, [
          {
            ...options,
            distDir,
            HtmlPlugin,
          },
        ]);

        const isRspack = api.context.bundlerType === 'rspack';
        chain
          .plugin(CHAIN_ID.PLUGIN.ASYNC_CHUNK_RETRY)
          .use(AsyncChunkRetryPlugin, [{ ...options, isRspack }]);
      },
    );
  },
});
