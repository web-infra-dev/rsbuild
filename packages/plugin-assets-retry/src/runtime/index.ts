// rsbuild/initial-chunk/retry
import type { CrossOrigin } from '@rsbuild/shared';
import type {
  AssetsRetryHookContext,
  PluginAssetsRetryOptions,
} from '../types';

interface ScriptElementAttributes {
  url: string;
  times: number;
  isAsync: boolean;
  crossOrigin?: CrossOrigin | boolean;
}

const TAG_TYPE: { [propName: string]: new () => HTMLElement } = {
  link: HTMLLinkElement,
  script: HTMLScriptElement,
  img: HTMLImageElement,
};
const TYPES = Object.keys(TAG_TYPE);

function findCurrentDomain(url: string, domainList: string[]) {
  let domain = '';
  for (let i = 0; i < domainList.length; i++) {
    if (url.indexOf(domainList[i]) !== -1) {
      domain = domainList[i];
      break;
    }
  }
  return domain || url;
}

function findNextDomain(url: string, domainList: string[]) {
  const currentDomain = findCurrentDomain(url, domainList);
  const index = domainList.indexOf(currentDomain);
  return domainList[(index + 1) % domainList.length] || url;
}

function getRequestUrl(element: HTMLElement) {
  if (
    element instanceof HTMLScriptElement ||
    element instanceof HTMLImageElement
  ) {
    return element.src;
  }
  if (element instanceof HTMLLinkElement) {
    return element.href;
  }
  return null;
}

const defaultConfig: PluginAssetsRetryOptions = {
  max: 3,
  type: TYPES,
  domain: [],
  crossOrigin: false,
};

function validateTargetInfo(
  config: PluginAssetsRetryOptions,
  e: Event,
): { target: HTMLElement; tagName: string; url: string } | false {
  const target: HTMLElement = e.target as HTMLElement;
  const tagName = target.tagName?.toLocaleLowerCase();
  const allowTags = config.type!;
  const url = getRequestUrl(target);
  if (
    !tagName ||
    allowTags.indexOf(tagName) === -1 ||
    !TAG_TYPE[tagName] ||
    !(target instanceof TAG_TYPE[tagName]) ||
    !url
  ) {
    return false;
  }
  return { target, tagName, url };
}

function createElement(
  origin: HTMLElement,
  attributes: ScriptElementAttributes,
): { element: HTMLElement; str: string } | undefined {
  const crossOrigin =
    attributes.crossOrigin === true ? 'anonymous' : attributes.crossOrigin;
  const crossOriginAttr = crossOrigin ? `crossorigin="${crossOrigin}"` : '';
  const retryTimesAttr = attributes.times
    ? `data-rsbuild-retry-times="${attributes.times}"`
    : '';
  const isAsyncAttr = attributes.isAsync ? 'data-rsbuild-async' : '';

  if (origin instanceof HTMLScriptElement) {
    const script = document.createElement('script');
    script.src = attributes.url;
    if (crossOrigin) {
      script.crossOrigin = crossOrigin;
    }
    if (attributes.times) {
      script.dataset.rsbuildRetryTimes = String(attributes.times);
    }
    if (attributes.isAsync) {
      script.dataset.rsbuildAsync = '';
    }
    return {
      element: script,
      str:
        // biome-ignore lint/style/useTemplate: use "</" + "script>" instead of script tag to avoid syntax error when inlining in html
        `<script src="${attributes.url}" ${crossOriginAttr} ${retryTimesAttr} ${isAsyncAttr}>` +
        '</' +
        'script>',
    };
  }
  if (origin instanceof HTMLLinkElement) {
    const link = document.createElement('link');
    link.rel = origin.rel || 'stylesheet';

    if (origin.as) {
      link.as = origin.as;
    }

    link.href = attributes.url;
    if (crossOrigin) {
      link.crossOrigin = crossOrigin;
    }
    if (attributes.times) {
      link.dataset.rsbuildRetryTimes = String(attributes.times);
    }
    return {
      element: link,
      str: `<link rel="${link.rel}" href="${
        attributes.url
      }" ${crossOriginAttr} ${retryTimesAttr} ${
        link.as ? `as="${link.as}"` : ''
      }></link>`,
    };
  }
}

function reloadElementResource(
  origin: HTMLElement,
  fresh: { element: HTMLElement; str: string },
  options: ScriptElementAttributes,
) {
  if (origin instanceof HTMLScriptElement) {
    if (options.isAsync) {
      document.body.appendChild(fresh.element);
    } else {
      document.write(fresh.str);
    }
  }

  if (origin instanceof HTMLLinkElement) {
    document.getElementsByTagName('head')[0].appendChild(fresh.element);
  }

  if (origin instanceof HTMLImageElement) {
    origin.src = options.url;
    origin.dataset.rsbuildRetryTimes = String(options.times);
  }
}

function retry(config: PluginAssetsRetryOptions, e: Event) {
  const targetInfo = validateTargetInfo(config, e);
  if (targetInfo === false) {
    return;
  }
  const { target, tagName, url } = targetInfo;
  // Filter by config.test and config.domain
  let tester = config.test;
  if (tester) {
    if (typeof tester === 'string') {
      const regexp = new RegExp(tester);
      tester = (str: string) => regexp.test(str);
    }

    if (typeof tester !== 'function' || !tester(url)) {
      return;
    }
  }

  const domain = findCurrentDomain(url, config.domain!);

  if (
    config.domain &&
    config.domain.length > 0 &&
    config.domain.indexOf(domain) === -1
  ) {
    return;
  }

  // If the retry times has exceeded the maximum, fail
  const existRetryTimes = Number(target.dataset.rsbuildRetryTimes) || 0;
  if (existRetryTimes === config.max!) {
    if (typeof config.onFail === 'function') {
      const context: AssetsRetryHookContext = {
        times: existRetryTimes,
        domain,
        url,
        tagName,
      };
      config.onFail(context);
    }
    return;
  }

  // Then, we will start to retry
  const nextDomain = findNextDomain(domain, config.domain!);

  const isAsync =
    Boolean(target.dataset.rsbuildAsync) ||
    (target as HTMLScriptElement).async ||
    (target as HTMLScriptElement).defer;

  const attributes: ScriptElementAttributes = {
    url: url.replace(domain, nextDomain),
    times: existRetryTimes + 1,
    crossOrigin: config.crossOrigin,
    isAsync,
  };
  const element = createElement(target, attributes)!;

  if (config.onRetry && typeof config.onRetry === 'function') {
    const context: AssetsRetryHookContext = {
      times: existRetryTimes,
      domain,
      url,
      tagName,
    };
    config.onRetry(context);
  }

  reloadElementResource(target, element, attributes);
}

function load(config: PluginAssetsRetryOptions, e: Event) {
  const targetInfo = validateTargetInfo(config, e);
  if (targetInfo === false) {
    return;
  }
  const { target, tagName, url } = targetInfo;
  const domain = findCurrentDomain(url, config.domain!);
  const retryTimes = Number(target.dataset.rsbuildRetryTimes) || 0;
  if (retryTimes === 0) {
    return;
  }
  if (typeof config.onSuccess === 'function') {
    const context: AssetsRetryHookContext = {
      times: retryTimes,
      domain,
      url,
      tagName,
    };
    config.onSuccess(context);
  }
}

function resourceMonitor(
  error: (e: Event) => void,
  success: (e: Event) => void,
) {
  if (typeof window !== 'undefined' && typeof window.document !== 'undefined') {
    document.addEventListener(
      'error',
      (e) => {
        if (e && e.target instanceof Element) {
          error(e);
        }
      },
      true,
    );

    document.addEventListener(
      'load',
      (e) => {
        if (e && e.target instanceof Element) {
          success(e);
        }
      },
      true,
    );
  }
}

// @ts-expect-error init is a global function, ignore ts(6133)
function init(options: PluginAssetsRetryOptions) {
  const config: PluginAssetsRetryOptions = {};

  for (const key in defaultConfig) {
    // @ts-ignore
    config[key] = defaultConfig[key];
  }

  for (const key in options) {
    // @ts-ignore
    config[key] = options[key];
  }

  // Normalize config
  if (!Array.isArray(config.type) || config.type.length === 0) {
    config.type = defaultConfig.type;
  }
  if (!Array.isArray(config.domain) || config.domain.length === 0) {
    config.domain = defaultConfig.domain;
  }

  if (Array.isArray(config.domain)) {
    config.domain = config.domain.filter(Boolean);
  }

  // Bind event in window
  try {
    resourceMonitor(
      (e: Event) => {
        try {
          retry(config, e);
        } catch (err) {
          console.error('retry error captured', err);
        }
      },
      (e: Event) => {
        try {
          load(config, e);
        } catch (err) {
          console.error('load error captured', err);
        }
      },
    );
  } catch (err) {
    console.error('monitor error captured', err);
  }
}
