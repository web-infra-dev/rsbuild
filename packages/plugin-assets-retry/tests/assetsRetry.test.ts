import { createStubRsbuild } from '../../compat/webpack/tests/helper';
import { pluginAssetsRetry } from '../src';

describe('plugin-assets-retry', () => {
  it('should add assets retry plugin', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginAssetsRetry()],
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it("should't add assets retry plugin when target is set to 'node'", async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginAssetsRetry()],
      rsbuildConfig: {
        output: {
          targets: ['node'],
        },
      },
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });
});
