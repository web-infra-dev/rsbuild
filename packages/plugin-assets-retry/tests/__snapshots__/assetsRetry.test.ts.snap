// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-assets-retry > should accept user custom options 1`] = `
{
  "plugins": [
    AssetsRetryPlugin {
      "HtmlPlugin": [Function],
      "distDir": "static/js",
      "inlineScript": true,
      "minify": true,
      "name": "AssetsRetryPlugin",
      "scriptPath": "",
    },
    AsyncChunkRetryPlugin {
      "name": "ASYNC_CHUNK_RETRY_PLUGIN",
      "options": {
        "crossOrigin": true,
        "domain": [
          "/http/localhost:3003",
          "/http/localhost:3002",
          "/http/localhost:3001",
          "/http/localhost:3000",
        ],
        "inlineScript": true,
        "isRspack": false,
        "max": 4,
        "minify": true,
        "onFail": [Function],
        "onRetry": [Function],
        "onSuccess": [Function],
        "test": [Function],
        "type": [
          "script",
          "link",
        ],
      },
      "runtimeOptions": {
        "domain": [
          "/http/localhost:3003",
          "/http/localhost:3002",
          "/http/localhost:3001",
          "/http/localhost:3000",
        ],
        "max": 4,
        "onFail": [Function],
        "onRetry": [Function],
        "onSuccess": [Function],
        "test": [Function],
      },
    },
  ],
}
`;

exports[`plugin-assets-retry > should add assets retry plugin 1`] = `
{
  "plugins": [
    AssetsRetryPlugin {
      "HtmlPlugin": [Function],
      "distDir": "static/js",
      "inlineScript": true,
      "minify": true,
      "name": "AssetsRetryPlugin",
      "scriptPath": "",
    },
    AsyncChunkRetryPlugin {
      "name": "ASYNC_CHUNK_RETRY_PLUGIN",
      "options": {
        "crossOrigin": false,
        "isRspack": false,
        "minify": true,
      },
      "runtimeOptions": {},
    },
  ],
}
`;

exports[`plugin-assets-retry > should't add assets retry plugin when target is set to 'node' 1`] = `{}`;
